### PR TITLE
Add Monitoring V2

### PR DESCRIPTION
**Pull request by aiyengar2**
Introduces two charts:
- `rancher-monitoring`: deploys Prometheus Operator + Prometheus Adapter + a set of PushProx exporters depending on the k8s distribution the user is using (i.e. RKE, k3s, kubeAdm, etc.) 
- `rancher-pushprox`: deploys a single PushProx exporter consisting of a Deployment + Service of a proxy, a DaemonSet + Service of clients, a ServiceMonitor, and some RBAC components (if using a service account to make authorized scrapes on k8s components)

All the changes from upstream Prometheus Operator have been documented in the `overlay/CHANGELOG.md` file that can be viewed in each commit to see the specific changes that each commit introduces to the chart.

### CODE REVIEW

File: docs/index.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ entries:
   canal:
   - apiVersion: v1
     appVersion: v3.13.3
-    created: "2020-06-22T19:23:35.619402732Z"
+    created: "2020-06-26T11:06:00.318061-07:00"
```
Comment (aiyengar2) [inline docs/index.yaml]: @StrongMonkey it seems like all of the timestamps in`docs/index.yaml` are being updated on running `make CHART=rancher-monitoring-v2 charts` due to https://github.com/rancher/dev-charts/blob/f091a152fbe8ece1a05a4cb8e800aa217d48f825/scripts/generate-charts#L22

Is this expected?

Comment (StrongMonkey) [reply docs/index.yaml]: Yes, I think this is a known issue. Don't know how to fix it yet.

**Comment by StrongMonkey** [comment]
@aiyengar2 There is no need to commit changes for docs/index.yaml and tar.gz files. CI will publish that. You will just need to commit patch file and package.yaml.

**Comment by aiyengar2** [comment]
> @aiyengar2 There is no need to commit changes for docs/index.yaml and tar.gz files. CI will publish that. You will just need to commit patch file and package.yaml.

Sounds good, removed those files from the commit!

File: packages/cattle-pushprox/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# cattle-pushprox
+
+A Rancher chart based on [PushProx](https://github.com/prometheus-community/PushProx) that sets up a deployment of the PushProx proxy and several PushProx clients on a Kubernetes cluster.
+
+Installs [cattle-pushprox](https://github.com/rancher/dev-charts/tree/master/packages/cattle-pushprox) to create a DaemonSet of clients that can access their host's network and register with a PushProx proxy Deployment. A [Prometheus Operator](https://github.com/coreos/prometheus-operator) ServiceMonitor CR is also included that is configured to scrape the metrics from each of the clients through the proxy.
+
+The clients and proxy are created based on of a Rancher fork of the Prometheus Community [PushProx](https://github.com/prometheus-community/PushProx) project.
+
+Using an instance of this chart is suitable for the following scenarios:
+- You need to scrape metrics from a port that should not be accessible outside of the host (i.e. scraping `etcd` metrics in a hardened cluster)
```
Comment (deniseschannon) [inline packages/cattle-pushprox/README.md]: Small nit, but these should be `e.g.`  (for example) not `i.e.` (that is)

**Review by pennyscissors — DISMISSED** [review]
lgtm

File: packages/cattle-pushprox/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+appVersion: 1.16.0
```
Comment (deniseschannon) [inline packages/cattle-pushprox/Chart.yaml]: Add the Rancher certified annotation to the chart. 

```
annotations:
  io.rancher.certified: rancher
```

File: packages/cattle-pushprox/README.md (hunk 2 above)
Comment (aiyengar2) [reply packages/cattle-pushprox/README.md]: good catch, updated in latest push

File: packages/cattle-pushprox/Chart.yaml (hunk 3 above)
Comment (aiyengar2) [reply packages/cattle-pushprox/Chart.yaml]: added to the latest push!

File: packages/cattle-pushprox/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+apiVersion: v1
+name: cattle-pushprox
+description: A Rancher chart based on PushProx that sets up a deployment of the PushProx proxy and several PushProx clients on a Kubernetes cluster.
```
Comment (dramich) [inline packages/cattle-pushprox/Chart.yaml]: Can this be clarified? What does 'several' mean?

File: packages/cattle-pushprox/README.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# cattle-pushprox
+
+A Rancher chart based on [PushProx](https://github.com/prometheus-community/PushProx) that sets up a deployment of the PushProx proxy and several PushProx clients on a Kubernetes cluster.
+
+Installs [cattle-pushprox](https://github.com/rancher/dev-charts/tree/master/packages/cattle-pushprox) to create a DaemonSet of clients that can access their host's network and register with a PushProx proxy Deployment. A [Prometheus Operator](https://github.com/coreos/prometheus-operator) ServiceMonitor CR is also included that is configured to scrape the metrics from each of the clients through the proxy.
+
+The clients and proxy are created based on of a Rancher fork of the Prometheus Community [PushProx](https://github.com/prometheus-community/PushProx) project.
```
Comment (dramich) [inline packages/cattle-pushprox/README.md]:
```suggestion
The clients and proxy are created based on a Rancher fork of the Prometheus Community [PushProx](https://github.com/prometheus-community/PushProx) project.
```

File: packages/cattle-pushprox/templates/pushprox-clients.yaml [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      tolerations: {{ toYaml .Values.clients.tolerations | nindent 6 }}
+{{- end }}
+      hostNetwork: true
+      dnsPolicy: ClusterFirstWithHostNet
+      {{- if and .Values.clients.https.enabled .Values.clients.https.useServiceAccountCredentials }}
+      serviceAccountName: {{ template "pushProxy.client.name" . }}
+      {{- end }}
+      containers:
+      - name: pushprox-client
+        image: arvindiyengar/pushprox-linux-amd64:add_flag_for_token_path
```
Comment (dramich) [inline packages/cattle-pushprox/templates/pushprox-clients.yaml]: Should be a rancher hosted image.

File: packages/cattle-pushprox/Chart.yaml (hunk 4 above)
Comment (aiyengar2) [reply packages/cattle-pushprox/Chart.yaml]: Sure, will change it to `a DaemonSet of PushProx clients`

File: packages/cattle-pushprox/templates/pushprox-clients.yaml [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+          echo "key: ${KEY_FILE_NAME}"
+          echo "cacert: ${CACERT_FILE_NAME}"
+
+          CERT_FILE_SOURCE=$(find /etc/source/ -type f -name "${CERT_FILE_NAME}" | sort -r | head -n 1)
+          KEY_FILE_SOURCE=$(find /etc/source/ -type f -name "${KEY_FILE_NAME}" | sort -r | head -n 1)
+          CACERT_FILE_SOURCE=$(find /etc/source/ -type f -name "${CACERT_FILE_NAME}" | sort -r | head -n 1)
+          
+          test -z ${CERT_FILE_SOURCE} && echo "Failed to find cacert file" && exit 1
+          test -z ${CERT_FILE_SOURCE} && echo "Failed to find cert file" && exit 1
+          test -z ${CERT_FILE_SOURCE} && echo "Failed to find key file" && exit 1
```
Comment (dramich) [inline packages/cattle-pushprox/templates/pushprox-clients.yaml]: These all do the same thing

File: packages/cattle-pushprox/README.md (hunk 5 above)
Comment (aiyengar2) [reply packages/cattle-pushprox/README.md]: Good catch, will be updated in the latest commit

File: packages/cattle-pushprox/templates/pushprox-clients.yaml (hunk 6 above)
Comment (aiyengar2) [reply packages/cattle-pushprox/templates/pushprox-clients.yaml]: This is pending merges from https://github.com/rancher/PushProx/pull/1 and https://github.com/rancher/PushProx/pull/2 for initializing the Rancher repo where the hosted image will be generated from

File: packages/cattle-pushprox/templates/pushprox-clients.yaml (hunk 7 above)
Comment (aiyengar2) [reply packages/cattle-pushprox/templates/pushprox-clients.yaml]: Good catch, this should be:
```suggestion
          test -z ${CERT_FILE_SOURCE} && echo "Failed to find cert file" && exit 1
          test -z ${KEY_FILE_SOURCE} && echo "Failed to find key file" && exit 1
          test -z ${CACERT_FILE_SOURCE} && echo "Failed to find cacert file" && exit 1
```

Making the change in the next commit

File: packages/rancher-monitoring/rancher-monitoring.patch [hunk 8]
```diff
@@ -0,0 +1,1352 @@
+diff -x '*.tgz' -x '*.lock' -uNr packages/rancher-monitoring/charts-original/Chart.yaml packages/rancher-monitoring/charts/Chart.yaml
+--- packages/rancher-monitoring/charts-original/Chart.yaml
++++ packages/rancher-monitoring/charts/Chart.yaml
+@@ -1,5 +1,9 @@
+ apiVersion: v1
+ appVersion: 0.38.1
++annotations:
++  catalog.rancher.io/certified: rancher
++  catalog.rancher.io/namespace: rancher-monitoring
```
Comment (deniseschannon) [inline packages/rancher-monitoring/rancher-monitoring.patch]: Let's update this to `monitoring-system`. UI has requested that all the namespaces end in `-system` to allow filtering.

**Review by deniseschannon — CHANGES_REQUESTED** [review]

File: packages/rancher-pushprox/Chart.yaml [hunk 9]
```diff
@@ -0,0 +1,23 @@
+apiVersion: v1
+appVersion: 1.16.0
+annotations:
```
Comment (deniseschannon) [inline packages/rancher-pushprox/Chart.yaml:4]: Unclear if these are needed at this level, but let's leave them in for now.

File: packages/rancher-monitoring/rancher-monitoring.patch [hunk 10]
```diff
@@ -0,0 +1,1352 @@
...
+--- packages/rancher-monitoring/charts-original/Chart.yaml
++++ packages/rancher-monitoring/charts/Chart.yaml
+@@ -1,5 +1,9 @@
+ apiVersion: v1
+ appVersion: 0.38.1
++annotations:
++  catalog.rancher.io/certified: rancher
++  catalog.rancher.io/namespace: monitoring-system
++  catalog.rancher.io/release-name: rancher-monitoring
+ description: Provides easy monitoring definitions for Kubernetes services, and deployment
```
Comment (deniseschannon) [inline packages/rancher-monitoring/rancher-monitoring.patch]: We should update the description to add in what other charts that we're adding as part of our version of the chart.

Comment (aiyengar2) [reply packages/rancher-monitoring/rancher-monitoring.patch]: Updated to `A Rancher chart that modifies the upstream Prometheus Operator chart, which provides easy monitoring definitions for Kubernetes services and the deployment and management of Prometheus instances, and enables Prometheus Adapter on a default Prometheus instance.`

File: packages/rancher-pushprox/Chart.yaml (hunk 9 above)
Comment (aiyengar2) [reply packages/rancher-pushprox/Chart.yaml:4]: I updated the appVersion to 0.1.0 to match the version of the upstream PushProx we are deploying:

https://github.com/rancher/PushProx/blob/753361c573410dc535028be56022962b9df3b9a6/VERSION#L1

File: packages/rancher-monitoring/rancher-monitoring.patch (hunk 8 above)
Comment (aiyengar2) [reply packages/rancher-monitoring/rancher-monitoring.patch]: Updated in the latest commit

**Review by deniseschannon — APPROVED** [review]
LGTM for packaging

**Review by dramich — APPROVED** [review]

**Pull request merged**